### PR TITLE
Bundler Package

### PR DESCRIPTION
**Pull request by talhamalik883**
This PR includes bundler package integration in modular sdk code

### CODE REVIEW

File: packages/bundler-rpc/src/interfaces/IBundler.interface.ts
Comment (tomarsachin2271) [inline packages/bundler-rpc/src/interfaces/IBundler.interface.ts:1]: Keep the interface name as IBundler.ts

File: packages/bundler-rpc/src/interfaces/IBundler.interface.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { SendUserOpResponse, getUserOpGasPricesResponse } from "../types/Bundler.types"
+import { ChainId,UserOperation  } from '@biconomy/core-types'
```
Comment (tomarsachin2271) [inline packages/bundler-rpc/src/interfaces/IBundler.interface.ts:2]: Linting error. Give a space after `ChainId,`

File: packages/bundler-rpc/src/types/Bundler.types.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+export type SendUserOpResponse = {
```
Comment (tomarsachin2271) [inline packages/bundler-rpc/src/types/Bundler.types.ts:1]: These types should return only the object mentioned in data field. code and message are returned from the API call, and it should be handled inside the sdk methods. 

If code is anything other than 2xx then sdk should throw an exception with proper message so dapps can handle it at their end.

Types name should be

UserOpResponse = {
    userOpHash: string,
    wait: <mention function type here that return Promise<TransactionResponse>>
}

UserOpGasFieldsResponse = {
        preVerificationGas: string,
        verificationGasLimit: string,
        callGasLimit: string,
        maxPriorityFeePerGas: string,
        maxFeePerGas: string
}

I removed the gasPrice field. Why there is need to gasPrice field? It's not used in creating UserOp

File: packages/bundler-rpc/package.json
[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/package.json:1]: resolved

File: packages/bundler-rpc/package.json [hunk 3]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "@biconomy/bundler-rpc",
+  "version": "2.0.0",
+  "description": "Biconomy Bundler Rpc to interact with any bundler",
+  "main": "./dist/src/index.js",
+  "typings": "./dist/src/index.d.ts",
+  "keywords": [
+    "Ethereum",
+    "Gnosis",
```
[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/package.json:9]: resolved

File: packages/bundler-rpc/package.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "@biconomy/bundler-rpc",
+  "version": "2.0.0",
+  "description": "Biconomy Bundler Rpc to interact with any bundler",
```
[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/package.json:4]: resolved

File: packages/bundler-rpc/src/Bundler.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import { IBundler } from "./interfaces/IBundler.interface";
+import { UserOperation, ChainId } from '@biconomy/core-types'
+import { SendUserOpResponse, getUserOpGasPricesResponse } from "./types/Bundler.types"
+import { resolveProperties } from 'ethers/lib/utils'
+import { deepHexlify } from '@biconomy/common'
+import { HttpMethod, sendRequest } from './utils/httpRequests'
+
+export class Bundler implements IBundler {
+
+    constructor(readonly bundlerUrl: string,
```
[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/src/Bundler.ts:10]: resolved

File: packages/bundler-rpc/src/Bundler.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import { IBundler } from "./interfaces/IBundler.interface";
+import { UserOperation, ChainId } from '@biconomy/core-types'
+import { SendUserOpResponse, getUserOpGasPricesResponse } from "./types/Bundler.types"
+import { resolveProperties } from 'ethers/lib/utils'
+import { deepHexlify } from '@biconomy/common'
+import { HttpMethod, sendRequest } from './utils/httpRequests'
+
+export class Bundler implements IBundler {
```
[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/src/Bundler.ts:8]: resolved

File: packages/bundler-rpc/src/Bundler.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        readonly dappAPIKey: string) {
+    }
+
+    /**
+     * 
+     * @param chainId 
+     * @description This function will fetch gasPrices from bundler
+     * @returns 
+     */
+    async getUserOpGasPrices(chainId: ChainId): Promise<getUserOpGasPricesResponse> {
```
[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/src/Bundler.ts:22]: inside function we are using sendRequest method and it throw exception whatever it gets from server

[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/src/Bundler.ts:22]: resolved

[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/src/Bundler.ts:22]: resolved

File: packages/bundler-rpc/src/Bundler.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                method: 'eth_getUserOpGasPrices',
+                params: [chainId],
+                id: 1234,
+                jsonrpc: '2.0'
+            }
+        })
+        return response
+    }
+    /**
+     *
```
[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/src/Bundler.ts:36]: have removed the dapp key as its going to be the part of url

File: packages/bundler-rpc/src/interfaces/IBundler.interface.ts (hunk 1 above)
Comment (talhamalik883) [reply packages/bundler-rpc/src/interfaces/IBundler.interface.ts:2]: resolved

File: packages/bundler-rpc/src/types/Bundler.types.ts (hunk 2 above)
Comment (talhamalik883) [reply packages/bundler-rpc/src/types/Bundler.types.ts:1]: code and message are removed and types name are updated. gasPrice filed is just of type1 transaction as bundler just send gasPrice for non 1155 blockchain so we assign gasPrice value to maxFeePerGas and maxPriorityFeePerGas in userOP

File: packages/bundler-rpc/src/Bundler.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+     * @returns 
+     */
+    async getUserOpGasPrices(chainId: ChainId): Promise<getUserOpGasPricesResponse> {
+        const response: any = await sendRequest({
+            url: `${this.bundlerUrl}`,
+            method: HttpMethod.Post,
+            body: {
+                method: 'eth_getUserOpGasPrices',
+                params: [chainId],
+                id: 1234,
```
[the remark this answers is not in the document]
Comment (talhamalik883) [reply packages/bundler-rpc/src/Bundler.ts]: not sure sachin. chirag wrote these calls before. I have added timestamp for it now

**Comment by talhamalik883** [comment]
Some of the changes that i did in bundler package in my latest work is also update in this branch

File: packages/bundler-rpc/src/types/Bundler.types.ts (hunk 2 above)
Comment (tomarsachin2271) [reply packages/bundler-rpc/src/types/Bundler.types.ts]: See the method in EntryPoint.sol as well

```
function getUserOpGasPrice(MemoryUserOp memory mUserOp) internal view returns (uint256) {
    unchecked {
        uint256 maxFeePerGas = mUserOp.maxFeePerGas;
        uint256 maxPriorityFeePerGas = mUserOp.maxPriorityFeePerGas;
        if (maxFeePerGas == maxPriorityFeePerGas) {
            //legacy mode (for networks that don't support basefee opcode)
            return maxFeePerGas;
        }
        return min(maxFeePerGas, maxPriorityFeePerGas + block.basefee);
    }
```

Comment (talhamalik883) [reply packages/bundler-rpc/src/types/Bundler.types.ts]: > See the method in EntryPoint.sol as well
> 
> ```
> function getUserOpGasPrice(MemoryUserOp memory mUserOp) internal view returns (uint256) {
>     unchecked {
>         uint256 maxFeePerGas = mUserOp.maxFeePerGas;
>         uint256 maxPriorityFeePerGas = mUserOp.maxPriorityFeePerGas;
>         if (maxFeePerGas == maxPriorityFeePerGas) {
>             //legacy mode (for networks that don't support basefee opcode)
>             return maxFeePerGas;
>         }
>         return min(maxFeePerGas, maxPriorityFeePerGas + block.basefee);
>     }
> ```

So bundler don't have to send gasPrice. Instead it should send same gasPrice value in both maxFeePerGas and maxPriorityFeePerGas

**Review by livingrockrises — COMMENTED** [review]
does this change also include forward scrapping?
seeing lot of files being deleted

**Review by livingrockrises — APPROVED** [review]
looks good

**Pull request merged**